core/arch/aarch64/emit_utils.c

```diff
@@ -205,24 +272,76 @@ exit_cti_reaches_target(dcontext_t *dcontext, fragment_t *f, linkstub_t *l,
 }
 
 void
-patch_stub(fragment_t *f, cache_pc stub_pc, cache_pc target_pc, bool hot_patch)
+patch_stub(fragment_t *f, cache_pc stub_pc, cache_pc target_pc, cache_pc target_prefix_pc,
+           bool hot_patch)
 {
+    /* Compute offset as unsigned, modulo arithmetic. */
+    ptr_uint_t off = (ptr_uint_t)target_pc - (ptr_uint_t)stub_pc;
+    if (off + 0x8000000 < 0x10000000) {
+        /* target_pc is a near fragment. We can get there with a B
+         * (OP_b, 26-bit signed immediate offset).
+         * i#1911: Patching arbitrary instructions to an unconditional branch
+         * is theoretically not sound. Architectural specifications do not
+         * guarantee safe behaviour or any bound on when the change will be
+         * visible to other process elements.
+         */
+        *(uint *)vmcode_get_writable_addr(stub_pc) =
+            (0x14000000 | (0x03ffffff & off >> 2));
+        if (hot_patch)
+            machine_cache_sync(stub_pc, stub_pc + 4, true);
+        return;
+    }
+    /* target_pc is a far fragment. We must use an indirect branch. Note that the indirect
+     * branch needs to be to the fragment prefix, as we need to restore the clobbered
+     * regs. */
     byte *target_pc_slot = get_target_pc_slot(f, stub_pc);
-    *(uint64 *)target_pc_slot = (ptr_uint_t)target_pc;
+    /* We set hot_patch to false as we are not modifying code. */
+    ATOMIC_8BYTE_ALIGNED_WRITE(target_pc_slot, (ptr_uint_t)target_prefix_pc,
+                               /*hot_patch=*/false);
+    return;
 }
 
 bool
 stub_is_patched(dcontext_t *dcontext, fragment_t *f, cache_pc stub_pc)
 {
-    uint *addr = (uint *)get_target_pc_slot(f, stub_pc);
-    return addr != (uint *)fcache_return_routine(dcontext);
+    uint enc;
+    ATOMIC_4BYTE_ALIGNED_READ(vmcode_get_writable_addr(stub_pc), &enc);
+    /* linked to intermediate fragment. */
+    if ((enc & 0x14000000) == 0x14000000)
+        return true;
+
+    ptr_uint_t target_pc;
+    ATOMIC_8BYTE_ALIGNED_READ(get_target_pc_slot(f, stub_pc), &target_pc);
+    /* linked to far fragment. */
+    return target_pc != (ptr_uint_t)fcache_return_routine(dcontext);
 }
 
 void
 unpatch_stub(dcontext_t *dcontext, fragment_t *f, cache_pc stub_pc, bool hot_patch)
 {
+    /* This implementation consists of two operations to unpatch, but at any given time
+     * only one patching strategy would be used. So, one of these two operations would
+     * really be a no-op. */
```

Comment:
Should we read the current state to figure out whether each one needs to be done?  Generally a read and conditional to skip a write is worthwhile, performance-wise.

Reply:
Thanks, I added a read to pick which write needs to be done.

Then I also found that unpatch_stub is called blindly in some cases. So I made both writes conditional on their respective conditions.

Reply:
Let me know if you think there are any atomicity concerns with the current implementation of `unpatch_stub`.

Reply:
I think it makes sense: cache sync for tweaking code (xref the separate threads on that but those will be dealt w/ separately as discussed), store-release for changing data.